Sources/SourceKitLSP/Swift/CodeActions/ApplyDeMorganLaw.swift

```diff
@@ -0,0 +1,730 @@
+    ///  1. The negated expression is a composite expression, and
+    ///  2. The original expression is not a single.
+    ///
+    /// - Note: A composite expression is one of `AsExprSyntax`, `AwaitExprSyntax`, `InfixOperatorExprSyntax`,
+    /// `IsExprSyntax`, or `TryExprSyntax`.
+    ///
+    /// An example of mapping is as follows,
+    /// ```swift
+    /// // original expression
+    /// ((!a || b is String))
+    /// // expression of interest
+    /// !a || b is String
+    /// !a /* negated to and mapped to */ a
+    /// b is String /* negated to */ Negation(b is String) /* mapped to */ !(b is String)
+    /// // negated expression of interest
+    /// a && !(b is String)
+    /// // mapped to
+    /// !((a && !(b is String)))
+    /// ```
+    func map(negation: Negation) -> Negated {
+      let negatedExpr =
+        switch self {
+        case .single(let tuple, _):
+          PrefixOperatorExprSyntax(
+            operator: negation.exprType.negationPrefix,
+            expression: tuple.map(complementExpr: negation.expr)
+          )
+        case .other:
+          PrefixOperatorExprSyntax(
+            operator: negation.exprType.negationPrefix,
+            expression: negation.expr.isComposite
+              ? ExprSyntax(TupleExprSyntax(onlyExpr: negation.expr)) : negation.expr
+          )
+        }
+      return (ExprSyntax(negatedExpr), .negation)
+    }
+
+    /// Substitutes the expression of interest with the negated expression and returns the result.
+    func map(negatedExpr: some ExprSyntaxProtocol) -> ExprSyntax {
+      switch self {
+      case .single(let tuple, _):
+        ExprSyntax(tuple.map(complementExpr: ExprSyntax(negatedExpr)))
+      case .other:
+        ExprSyntax(negatedExpr)
+      }
+    }
+  }
+
+  /// Represents a negation expression with `expr` being its only sub-expression.
+  private struct Negation {
+    var exprType: ExprType
+    var expr: ExprSyntax
+
+    init(exprType: ExprType, expr: some ExprSyntaxProtocol) {
+      self.exprType = exprType
+      self.expr = ExprSyntax(expr)
+    }
+
+    /// If `exprType` is `nil`, this indicates the negation expression is at the root-level and `exprType` will be
+    /// inferred from the prefix of `prefixExpr`. Otherwise `prefixExpr` will be checked for its prefix's agreement
+    /// with the given `exprType`.
+    init?(_ prefixExpr: PrefixOperatorExprSyntax, exprType: ExprType? = nil) {
+      guard let localExprType = ExprType(operator: prefixExpr.operator) else {
+        return nil
+      }
+      if let exprType, exprType != localExprType {
+        return nil
+      }
+      self.init(
+        exprType: localExprType,
+        expr: prefixExpr.expression.with(
+          \.leadingTrivia,
+          prefixExpr.leadingTrivia + prefixExpr.expression.leadingTrivia
+        )
+      )
+    }
+  }
+
+  /// Represents the negated expression of an instance of propositions.
+  private struct NegatedPropositions {
```

Comment:
Do we need the `NegatedPropositions` type at all? It seems like all we ever do with it is to pass it into `Expr.map`, so instead of having the `NegatedPropositions` type, could we just have a function that returns the negated `Expr`?

Reply:
We need to tell from negated propositions the type of its operator to determine whether we need to add extra parentheses to preserve precedence order. We can turn `operator` into a computed property though.

Reply:
Couldn’t you do that by running the negation and then when realizing that you might need to add operators, get the transformed operator using something like `expr.as(BinaryExprSyntax.self)?.operator`? That way we wouldn’t have to pass around redundant state.